### PR TITLE
Implemented HSM layer for PCS

### PR DESCRIPTION
**Pull request by mpkelly-hpe**
### Summary and Scope

Implemented HSM layer for PCS.  Improved unit testing process
to do full discovery of specified components.

### Issues and Related PRs

* Resolves CASMHMS-5390

### Testing

Tested on:

* Local build machine
* wasp

On local build machine, unit testing is done in a docker composed environment, so
real HSM, RTS, etc. are used.  Unit tests exercise all the new HSM layer functions.
All functions work as expected.

On wasp, ginned up a test application using 'main.go' in the PCS repo, causing it
to call basically the same functions as the unit testing does.  This is overkill,
but wanted to just be certain.

### Risks and Mitigations

No risk currently, not in use anywhere.

### CODE REVIEW

**Review by alvarez3-hpe — APPROVED** [review]

File: internal/hsm/interface.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	HSMGlobals HSM_GLOBALS
+}
+
+type HSMProvider interface {
+	Init(glb *HSM_GLOBALS) error
+	Ping() error
+	ReserveComponents(compList []ReservationData) ([]*ReservationData,error)
+	ReleaseComponents(compList []ReservationData) ([]*ReservationData,error)
+	CheckDeputyKeys(comp []ReservationData) error
+	FillComponentEndpointData(hd map[string]*HsmData) []error
```
Comment (alvarez3-hpe) [inline internal/hsm/interface.go:39]: '[]error' is the incorrect return type for FillComponentEndpointData() in state_manager.go. It should be 'error'.

Comment (mpkelly-hpe) [reply internal/hsm/interface.go]: Good catch.   Fixed.

Comment (nieuwsma) [reply internal/hsm/interface.go]: The reason why it was []error is because its an array of ComponentEndpoint data. Im not sure it should be changed to a singleton.

Comment (alvarez3-hpe) [reply internal/hsm/interface.go]: The function itself in state_manager.go is a singleton. This was just being changed here to match that.

As far as singleton vs array goes, the function isn't acting individually on each ComponentEndpoint. I think a single error is sufficient to say whether or not it successfully got info back from HSM.

File: internal/hsm/models.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+package hsm
+
+import (
+	"github.com/sirupsen/logrus"
+	"github.com/Cray-HPE/hms-certs/pkg/hms_certs"
+	"github.com/Cray-HPE/hms-base"
```
Comment (nieuwsma) [inline internal/hsm/models.go:28]: what do you need from hms-base? can you get it from hms-xname?

File: scripts/wait-for-discovery.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
Comment (nieuwsma) [inline scripts/wait-for-discovery.sh:1]: what is this script for?  is it used for the docker-compose environment?

File: docker-compose.developer.full.yaml [hunk 4]
```diff
@@ -172,8 +172,6 @@ services:
       - CRAY_VAULT_JWT_FILE=configs/token
       - CRAY_VAULT_ROLE_FILE=configs/namespace
       - CRAY_VAULT_AUTH_PATH=auth/token/create
-      - HTTPS_CERT=configs/rts.crt
```
Comment (nieuwsma) [inline docker-compose.developer.full.yaml:175]: why are these removed?

File: scripts/wait-for-discovery.sh (hunk 3 above)
Comment (mpkelly-hpe) [reply scripts/wait-for-discovery.sh:1]: Yes.  In testing the HSM layer I need discovery to be complete before running the tests.   I was originally doing this in the HSM layer test code, but would guess over time that other tests will have the same requirement.  Thus, this script causes the test to wait until HSM discovery is done before running any of the tests.

File: docker-compose.developer.full.yaml (hunk 4 above)
Comment (mpkelly-hpe) [reply docker-compose.developer.full.yaml:175]: I took them out when I thought they weren't needed.   I added them back in for the unit test yaml file -- I will put these back in.

Comment (nieuwsma) [reply docker-compose.developer.full.yaml:175]: can you look at dynamically generating: https://github.com/Cray-HPE/hms-tools/blob/master/hwval/gen-certs.sh

File: internal/hsm/models.go (hunk 2 above)
Comment (mpkelly-hpe) [reply internal/hsm/models.go:28]: I need the HSM-style  Component type.  It does not exist in hms-xname.

File: docker-compose.developer.full.yaml (hunk 4 above)
Comment (mpkelly-hpe) [reply docker-compose.developer.full.yaml:175]: Will try.

Comment (mpkelly-hpe) [reply docker-compose.developer.full.yaml:175]: OK, made the .crt and .key files "ephemeral".

File: Dockerfile.unittesting.Dockerfile [hunk 5]
```diff
@@ -26,7 +26,10 @@ FROM artifactory.algol60.net/docker.io/library/golang:1.16-alpine
 
 RUN set -ex \
     && apk -U upgrade \
-    && apk add build-base
+    && apk add build-base \
+    && apk add bash \
```
Comment (nieuwsma) [inline Dockerfile.unittesting.Dockerfile:30]: why are you adding this into the unit testing?

**Review by nieuwsma — CHANGES_REQUESTED** [review]

**Review by mjendrysik-hpe — APPROVED** [review]

File: Dockerfile.unittesting.Dockerfile (hunk 5 above)
Comment (mpkelly-hpe) [reply Dockerfile.unittesting.Dockerfile:30]: All of these additions are there in order to run the wait-for-discovery.sh script.   I wanted bash so
the script would be easier to write.   curl and jq are used to fetch info from HSM and parse it in 
order to see if discovery is complete.

File: internal/hsm/hsm_test.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+	"testing"
+	"github.com/Cray-HPE/hms-certs/pkg/hms_certs"
+	"github.com/Cray-HPE/hms-xname/xnametypes"
+	"github.com/Cray-HPE/hms-base"
+	"github.com/sirupsen/logrus"
+)
+
+var glogger = logrus.New()
+
+func TestInit(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mpkelly-hpe) [reply internal/hsm/hsm_test.go]: Done.
